cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -72,97 +59,78 @@ ElementDofLayout CoordinateElement::create_dof_layout() const
-std::array<double, 3>
-CoordinateElement::x0(const xt::xtensor<double, 2>& cell_geometry)
-{
-  std::array<double, 3> x0 = {0, 0, 0};
-  for (std::size_t i = 0; i < cell_geometry.shape(1); ++i)
-    x0[i] += cell_geometry(0, i);
-  return x0;
-}
-//-----------------------------------------------------------------------------
-void CoordinateElement::pull_back_affine(xt::xtensor<double, 2>& X,
-                                         const xt::xtensor<double, 2>& K,
-                                         const std::array<double, 3>& x0,
-                                         const xt::xtensor<double, 2>& x)
-{
-  assert(X.shape(0) == x.shape(0));
-  assert(X.shape(1) == K.shape(0));
-  assert(x.shape(1) == K.shape(1));
-
-  // Calculate X for each point
-  X.fill(0.0);
-  for (std::size_t p = 0; p < x.shape(0); ++p)
-    for (std::size_t i = 0; i < K.shape(0); ++i)
-      for (std::size_t j = 0; j < K.shape(1); ++j)
-        X(p, i) += K(i, j) * (x(p, j) - x0[j]);
-}
-//-----------------------------------------------------------------------------
-void CoordinateElement::pull_back_nonaffine(
-    xt::xtensor<double, 2>& X, const xt::xtensor<double, 2>& x,
-    const xt::xtensor<double, 2>& cell_geometry, double tol, int maxit) const
+void CoordinateElement::pull_back_nonaffine(mdspan2_t X, cmdspan2_t x,
+                                            cmdspan2_t cell_geometry,
+                                            double tol, int maxit) const
 {
   // Number of points
-  std::size_t num_points = x.shape(0);
+  std::size_t num_points = x.extent(0);
   if (num_points == 0)
     return;
 
   const std::size_t tdim = mesh::cell_dim(this->cell_shape());
-  const std::size_t gdim = x.shape(1);
-  const std::size_t num_xnodes = cell_geometry.shape(0);
-  assert(cell_geometry.shape(1) == gdim);
-  assert(X.shape(0) == num_points);
-  assert(X.shape(1) == tdim);
-
-  xt::xtensor<double, 2> dphi({tdim, num_xnodes});
-  xt::xtensor<double, 2> Xk({1, tdim});
+  const std::size_t gdim = x.extent(1);
+  const std::size_t num_xnodes = cell_geometry.extent(0);
+  assert(cell_geometry.extent(1) == gdim);
+  assert(X.extent(0) == num_points);
+  assert(X.extent(1) == tdim);
+
+  std::vector<double> dphi_b(tdim * num_xnodes);
+  mdspan2_t dphi(dphi_b.data(), tdim, num_xnodes);
+
+  std::vector<double> Xk_b(tdim);
+  mdspan2_t Xk(Xk_b.data(), 1, tdim);
+
   std::array<double, 3> xk = {0, 0, 0};
-  xt::xtensor<double, 1> dX = xt::empty<double>({tdim});
-  xt::xtensor<double, 2> J({gdim, tdim});
-  xt::xtensor<double, 2> K({tdim, gdim});
-  xt::xtensor<double, 4> basis(_element->tabulate_shape(1, 1));
+
+  std::vector<double> dX(tdim);
+
+  std::vector<double> J_b(gdim * tdim);
+  mdspan2_t J(J_b.data(), gdim, tdim);
+
+  std::vector<double> K_b(tdim * gdim);
+  mdspan2_t K(K_b.data(), tdim, gdim);
+
+  namespace stdex = std::experimental;
+  using mdspan4_t = stdex::mdspan<double, stdex::dextents<std::size_t, 4>>;
```

Comment:
Instead of redeclaring these span types all over the place, could we use those in 
https://github.com/FEniCS/basix/blob/3844e0f52e34106a9f23a2298a00224d816466e5/cpp/basix/finite-element.h#L25-L48
for all double spans?

Reply:
The `impl` namespace implies private/internal. We shouldn't use `impl` from other libraries.